### ISSUE TITLE
feat(nn): SIREN — sinusoidal representation network layer + composite

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#88]
## Summary

Add SIREN (Sinusoidal Representation Networks) to `pyrox.nn` — the single-layer primitive `SirenDense`, the composite network `SIREN`, and a Bayesian variant `BayesianSIREN`. SIREN is the canonical implicit neural representation (INR) backbone: an MLP with `sin` activations and a carefully derived three-regime initialization that preserves activation distribution across depth. It's the direct competitor to MFN (#87) as an INR basis, and the natural target of FiLM conditioning (#86) for amortized / conditional fields.

Landing this gives us (a) the standard INR baseline every other INR paper benchmarks against, (b) a SIREN-backed `BayesianNeuralField` variant, (c) the ingredients for modulated-SIREN (Mehta et al. 2021) once FiLM + SIREN both exist.

## Background

SIREN (Sitzmann, Martel, Bergman, Lindell, Wetzstein — NeurIPS 2020) replaces ReLU/GELU with `sin` and prescribes an initialization scheme that keeps the pre-activation variance stable at each layer. The architecture is:

$$
z_1 = \sin\bigl(\omega_0 (W_0 x + b_0)\bigr),
\quad
z_{i+1} = \sin\bigl(\omega (W_i z_i + b_i)\bigr),
\quad
y = W_L z_L + b_L.
$$

The three-regime init (Sitzmann §3.2, Theorem 1) is the whole point:

- **First layer**: $W_0 \sim \mathcal{U}(-1/d_{\text{in}},\, 1/d_{\text{in}})$ — bounds the first-layer pre-activation to $[-1, 1]$ so $\sin(\omega_0 \cdot)$ operates over exactly one period.
- **Hidden layer**: $W_i \sim \mathcal{U}\bigl(-\sqrt{6/d_h}/\omega,\, \sqrt{6/d_h}/\omega\bigr)$ — Sitzmann's derivation: under activations $\approx \mathcal{N}(0,1)$, this keeps post-sin activations $\approx \mathcal{N}(0,1)$.
- **Last layer**: $W_L \sim \mathcal{U}\bigl(-\sqrt{6/d_h},\, \sqrt{6/d_h}\bigr)$ — standard Glorot, no activation.

Typical choice: $\omega_0 = 30$ for image / high-frequency INR tasks; $\omega = 30$ in hidden layers (Sitzmann's default). Lower $\omega$ for smoother signals.

Why this matters for pyrox:

- **The INR baseline everyone compares to.** Without SIREN we can't reproduce or benchmark against any INR paper.
- **Natural base for FiLM conditioning.** Modulated SIREN (Mehta et al. 2021) is the standard conditional-INR recipe; it needs both SIREN and FiLM in the same library.
- **Bayesian INR is under-explored.** `BayesianSIREN` with priors on `W, b` is a research-relevant primitive that has no off-the-shelf JAX implementation.
- **Complements MFN (#87), not duplicates.** SIREN is a *deep composition of nonlinearities*; MFN is a *product of filters*. Different inductive biases, different empirical profiles — users should have both.

### References

- Sitzmann, Martel, Bergman, Lindell, Wetzstein. **"Implicit Neural Representations with Periodic Activation Functions."** NeurIPS 2020. [arXiv:2006.09661](https://arxiv.org/abs/2006.09661)
- Project page + reference PyTorch impl: [vsitzmann.github.io/siren](https://vsitzmann.github.io/siren/), [github.com/vsitzmann/siren](https://github.com/vsitzmann/siren)
- Mehta, Gharbi, Barnes, Shechtman, Ramamoorthi, Chandraker. **"Modulated Periodic Activations for Generalizable Local Functional Representations."** ICCV 2021. [arXiv:2104.03960](https://arxiv.org/abs/2104.03960) — FiLM-over-SIREN; unlocked once #86 + this issue land.
- Tancik et al. **"Fourier Features Let Networks Learn High Frequency Functions."** NeurIPS 2020. [arXiv:2006.10739](https://arxiv.org/abs/2006.10739) — the alternative framing (fixed Fourier features + ReLU MLP) that SIREN subsumes.
- Existing reference port in our own stack: [`jejjohnson/keras-nerfs/src/knerf/_src/layers/siren.py`](https://github.com/jejjohnson/keras-nerfs/blob/main/src/knerf/_src/layers/siren.py) — same author, Keras implementation to cross-check init against.

## Proposed API

All public classes land in `src/pyrox/nn/_layers.py` and re-export from `pyrox.nn`.

### Single layer

```python
class SirenDense(PyroxModule):
    """Sine-activated dense layer: ``y = sin(ω · (W x + b))`` or ``y = W x + b``.

    Three init regimes (Sitzmann 2020 Theorem 1):

    - ``layer_type="first"``:  ``W ~ U(-1/d_in, 1/d_in)``, sin activation ON.
    - ``layer_type="hidden"``: ``W ~ U(-sqrt(6/d_in)/ω, sqrt(6/d_in)/ω)``, sin ON.
    - ``layer_type="last"``:   ``W ~ U(-sqrt(6/d_in), sqrt(6/d_in))``, sin OFF.

    The ``c`` parameter in the original paper is fixed at 6 (from the
    U(-a, a) variance formula ``a² · 2/12 · 2 = a²/3`` matched to ``1/d_in``).
    We expose it for completeness but default to 6.
    """
    W: Float[Array, "in out"]
    b: Float[Array, " out"]
    omega: float
    in_features: int
    out_features: int
    layer_type: Literal["first", "hidden", "last"]
    c: float = 6.0

    @classmethod
    def init(
        cls, in_features: int, out_features: int,
        *, key: PRNGKeyArray,
        omega: float = 30.0,
        layer_type: Literal["first", "hidden", "last"] = "hidden",
        c: float = 6.0,
        pyrox_name: str | None = None,
    ) -> "SirenDense": ...

    @pyrox_method
    def __call__(self, x: Float[Array, "N D_in"]) -> Float[Array, "N D_out"]: ...
```

Semantics:

- `omega` is applied *inside* the sine: `sin(omega * (W x + b))`. It scales the frequency content, not the weight init directly — the init formula cancels it analytically so hidden-layer activations stay $\approx \mathcal{N}(0, 1)$.
- `b` is initialized with `U(-1/sqrt(d_in), 1/sqrt(d_in))` for all regimes (matches the `keras-nerfs` reference and Sitzmann's official PyTorch impl).
- `layer_type="last"` is the *pre-readout* convention: no sine, standard linear. Users who want a tanh or sigmoid output layer can compose externally.

### Composite network

```python
class SIREN(PyroxModule):
    """Multi-layer SIREN network with proper first/hidden/last init.

    Topology:
        z_1 = SirenDense(in_features, hidden_features, layer_type="first",
                         omega=first_omega)(x)
        z_{i+1} = SirenDense(hidden_features, hidden_features,
                             layer_type="hidden", omega=hidden_omega)(z_i)
        y = SirenDense(hidden_features, out_features, layer_type="last")(z_L)
    """
    layers: list[SirenDense]            # length depth; first + (depth-2) hidden + last
    in_features: int
    hidden_features: int
    out_features: int
    depth: int                          # total layers, including readout; depth >= 2
    first_omega: float
    hidden_omega: float

    @classmethod
    def init(
        cls, in_features: int, hidden_features: int, out_features: int,
        *, depth: int, key: PRNGKeyArray,
        first_omega: float = 30.0, hidden_omega: float = 30.0,
        c: float = 6.0,
        pyrox_name: str | None = None,
    ) -> "SIREN": ...

    @pyrox_method
    def __call__(self, x: Float[Array, "N D_in"]) -> Float[Array, "N D_out"]: ...
```

`depth` counts *all* layers including readout. `depth=2` = first + last (no hidden); `depth=5` = first + 3 hidden + last. Must be ≥ 2.

### Bayesian variant

```python
class BayesianSIREN(SIREN):
    """SIREN with priors on all layer weights.

    Registers sites ``{scope}.layer_{i}.W`` and ``{scope}.layer_{i}.b``
    via ``pyrox_sample`` for each of the ``depth`` layers. Default
    prior is ``Normal(0, prior_std)`` on weights and biases. The
    SIREN-specific init stddevs are ignored at sample time (the prior
    takes over), but the ``first/hidden/last`` scale is still applied
    as a *prior rescaling*: W_i ~ N(0, prior_std · init_scale_i²) so
    the Sitzmann-regime structure is preserved in expectation.

    This hybrid — Sitzmann-scaled Normal priors — is the principled
    Bayesian analog of SIREN's init. Users who want a flat ``N(0, 1)``
    prior on every layer should subclass.
    """
    prior_std: float = 1.0
```

The prior-rescaling detail is the load-bearing design choice: a flat $\mathcal{N}(0, 1)$ prior would immediately break the Sitzmann variance preservation and turn hidden activations into saturated sine regions. Scaling the prior by the init stddev keeps the Bayesian version statistically equivalent to the deterministic SIREN at initialization.

### No pure-JAX escape hatch

Unlike FiLM's `apply_film` or MFN's `mfn_forward`, there's no meaningful "SIREN without the module" — the whole point is the chained layers with regime-specific init. Users who want a one-off sine activation can use `jnp.sin` directly.

## Integration points

- **FiLM (#86)**: `SIREN → FiLM → SIREN → FiLM → ...` reproduces modulated SIREN (Mehta et al. 2021). Explicitly enabled by this issue + #86.
- **MFN (#87)**: positioned as the *alternative* INR backbone. `SIREN` and `GaborNet` should have a cross-linked docs section explaining the tradeoff (deep-composition vs product-of-filters).
- **`BayesianNeuralField`** (bayesnf port): `SIREN` is a drop-in MLP replacement. The internal `eqx.nn.Linear` + activation chain becomes `SIREN(...)`.
- **HSGP / VFF / RFF**: SIREN's first layer can be interpreted as a learnable RFF — worth a short docstring callout, not a shared implementation.
- **`GPPrior` / NN-GP kernels** (future): the infinite-width limit of a SIREN has a known kernel structure (Yuce et al. 2022, "A Structured Dictionary Perspective on Implicit Neural Representations"). Not in scope here but worth flagging for future theoretical work.

### File layout

- **Edit** `src/pyrox/nn/_layers.py`: add `SirenDense`, `SIREN`, `BayesianSIREN`. Order: primitive, composite, Bayesian variant.
- **Edit** `src/pyrox/nn/__init__.py`: re-export the three new public names.
- **New** `tests/nn/test_siren.py`: unit + property + correctness tests.
- **New** `docs/api/nn/siren.md`: mkdocstrings stubs with the three-regime init table in the page body.
- **Edit** `mkdocs.yml`: nav entry under `nn/`.

### Init (load-bearing — follow Sitzmann §3.2 exactly)

```python
def _siren_W_limit(layer_type, in_features, omega, c=6.0):
    if layer_type == "first":
        return 1.0 / in_features
    if layer_type == "hidden":
        return jnp.sqrt(c / in_features) / omega
    if layer_type == "last":
        return jnp.sqrt(c / in_features)
    raise ValueError(f"unknown layer_type {layer_type!r}")
```

Three implementation-level gotchas to call out in the PR:

1. **`omega` appears in the init denominator for hidden layers only.** Dropping it turns SIREN into a generic sine-MLP that saturates within 2-3 layers. Test #4 below catches this.
2. **`b` init is the same across regimes.** `U(-1/sqrt(in), 1/sqrt(in))`. Do not try to rescale by omega — the Sitzmann paper uses the default equinox / PyTorch bias init.
3. **Sine applies *after* the linear + bias, but *before* any further reshape.** The `omega` multiplier goes inside the sine on the pre-activation `(W x + b)`. Applying it to `W` before the matmul gives numerically different results under mixed precision.

### Forward pass — einops discipline

```python
@pyrox_method
def __call__(self, x):
    z = x
    for layer in self.layers:
        z = layer(z)
    return z
```

Single linear matmul per layer, followed by conditional sine. No einops needed beyond what `eqx.nn.Linear` already does internally.

### Handling `x.ndim == 1`

`atleast_2d` → run → squeeze. Same pattern as FiLM (#86) and MFN (#87).

### Type annotations

```python
from jaxtyping import Array, Float, PRNGKeyArray
from typing import Literal
```

Per-file ruff ignore `"src/pyrox/nn/**" = ["F722"]` already in effect. No new config.

## Tests — `tests/nn/test_siren.py`

All tests use existing numpyro/equinox fixtures for consistency with `test_spectral_layers.py`.

1. **`test_siren_dense_output_shape`** — `SirenDense(in=3, out=16, layer_type="hidden")` on `x: (5, 3)` → `(5, 16)`.
2. **`test_siren_dense_activation_regimes`** — `"first"` and `"hidden"` apply `sin`; `"last"` does not. Verify by constructing identical `W, b` across three instances and checking output equality vs. `jnp.sin(omega * (W x + b))` / `W x + b`.
3. **`test_siren_composite_output_shape`** — `SIREN(in=2, hidden=32, out=4, depth=5)` on `x: (10, 2)` → `(10, 4)`.
4. **`test_siren_activation_variance_preserved`** — Sitzmann Theorem 1 smoke test: with `x ~ N(0, 1)` of shape `(4096, 64)`, a depth-5 SIREN's hidden-layer activations have variance within `[0.3, 3.0]` at every layer. This is the **load-bearing correctness anchor** — if init is wrong, variance explodes or collapses within 2 layers.
5. **`test_siren_jits`** — forward under `jax.jit` returns finite values.
6. **`test_siren_gradient_flows`** — `jax.grad` of a scalar loss w.r.t. the full SIREN PyTree has finite non-zero gradient on every `W` and `b` leaf.
7. **`test_siren_fits_1d_sinusoid`** — fit a depth-3 SIREN (hidden=64) to `y = sin(2x)` on `x ∈ [-π, π]`, 256 points, 500 Adam steps. Assert final MSE < 1e-3. The tightest we can make this without a slow test — SIREN fits sinusoids *trivially* so the bound is safe, but if init regresses the loss will be 100× higher. Catches the whole class of silent-regression bugs.
8. **`test_bayesian_siren_registers_sites`** — traced forward registers exactly `2 · depth` sample sites (`W` + `b` per layer). Assert names and count.
9. **`test_bayesian_siren_prior_respects_init_scale`** — at init, sampled `W` stddev per layer matches `prior_std · init_scale_i` within 20% Monte-Carlo tolerance (32 samples, 512 hidden). Catches the prior-rescaling detail.
10. **`test_siren_rejects_invalid_depth`** — `depth=1` and `depth=0` raise `ValueError` containing `"depth"`.
11. **`test_siren_rejects_invalid_layer_type`** — `SirenDense.init(layer_type="mystery")` raises `ValueError` with message containing `"layer_type"`.
12. **`test_siren_first_omega_differs_from_hidden`** — `SIREN(first_omega=60, hidden_omega=30)` → assert `siren.layers[0].omega == 60` and `siren.layers[1].omega == 30`. Verifies the two-omega wiring.

Keep tests fast. Test #7 is the only one that trains — 500 Adam steps at hidden=64 is ~1 second on CPU. No notebook-style eval loops.

## Docs

- mkdocstrings page `docs/api/nn/siren.md`, one `:::` block per public class.
- Page body includes (a) the three-regime init table, (b) an explicit callout that `SIREN` and `GaborNet` (#87) are alternative INR backbones, (c) a 6-line usage example.
- Cross-link to FiLM (#86) docs noting that modulated-SIREN is enabled once both land.
- **No notebook** in this issue — add a SIREN section to the planned `bayesian_neural_fields.py` notebook in a follow-up.

## Acceptance criteria

- [ ] `pyrox.nn.SirenDense`, `pyrox.nn.SIREN`, `pyrox.nn.BayesianSIREN` are public and documented.
- [ ] All 12 tests in `tests/nn/test_siren.py` pass locally and under CI.
- [ ] `uv run pytest -q` green.
- [ ] `uv run --group lint ruff check .` green (entire repo).
- [ ] `uv run --group lint ruff format --check .` green (entire repo).
- [ ] `uv run --group typecheck ty check src/pyrox` green.
- [ ] **`test_siren_activation_variance_preserved`** passes — this is the non-negotiable correctness check that the three-regime init is implemented correctly.
- [ ] **`test_siren_fits_1d_sinusoid`** achieves MSE < 1e-3 — guards against the whole class of "init subtly broken" regressions.

## Out of scope (explicit)

- **FiLM-over-SIREN (modulated SIREN).** Unlocked once both #86 and this land; separate integration issue.
- **SIREN + MFN hybrids.** Niche; users compose by hand.
- **Learnable `omega`.** Not in Sitzmann's original recipe; empirical results mixed. Out of scope.
- **Complex-valued SIREN variants** (WIRE / COIN). Separate issue.
- **Pruning / distillation.** Separate topic.
- **Notebook.** Defer to the bayesian_neural_fields notebook follow-up.

## Estimated effort

Small-to-medium. `SirenDense` is ~40 lines; `SIREN` composite is ~50 lines; `BayesianSIREN` is ~30 lines; tests ~200 lines; docs ~50 lines. Single PR, no new dependencies. Main risk: getting any of the three init regimes wrong — tests #4 and #7 are the combined guardrail.
